tenants/tenants_test.go

```diff
@@ -0,0 +1,86 @@
+package tenants
+
+import (
+	"net/url"
+	"testing"
+	"time"
+
+	prometheus "github.com/prometheus/client_model/go"
+)
+
+// Tests if retries are working for tenant registrations.
+// To test this, a tenant with wrong configuration needs to be present at /test/conf/tenants.yaml.
+func TestListenAndServeTenants(t *testing.T) {
+	cfg := config{}
+
+	cfg.server.listen = "0.0.0.0:8443"
+	cfg.server.listenInternal = "0.0.0.0:8448"
+
+	u, _ := url.ParseRequestURI("http://127.0.0.1:3100")
+	cfg.logs.readEndpoint = u
+
+	u, _ = url.ParseRequestURI("http://127.0.0.1:3100")
+	cfg.logs.writeEndpoint = u
+
+	u, _ = url.ParseRequestURI("http://127.0.0.1:9091")
+	cfg.metrics.readEndpoint = u
+
+	u, _ = url.ParseRequestURI("http://127.0.0.1:19291")
+	cfg.metrics.writeEndpoint = u
+
+	cfg.rbacConfigPath = "../test/config/rbac.yaml"
+	cfg.tenantsConfigPath = "../test/config/tenants.yaml"
+	cfg.server.healthcheckURL = "https://127.0.0.1:8443"
+	cfg.tls.reloadInterval = time.Minute
+	cfg.middleware.concurrentRequestLimit = 1000
+	cfg.middleware.backLogLimitConcurrentRequests = 0
+	cfg.logLevel = "debug"
+
+	// command line configuration is now setup, populate tenantsConfig struct from commadline configuration.
+	tCfg := loadTenantConfigs(&cfg)
+
+	// onboard teants in a goroutine and watch retry metrics to go up as the tenant registration fails for
+	// a tenant with wrong configuration at ./test/conf/tenants.yaml.
```

Comment:
I was wondering whether we really need a wrongly-configured tenant for this test or if we could structure the test differently. Since we're not running Dex as a part of the test, even for 'correctly' configured tenants we would see increase in retry attempts for any tenant, since they won't be able to reach Dex.

Reply:
I see that we use Dex as part of CI tests.